### PR TITLE
feat: add Hear from an Artsy Advisor in saved search flow, behind ff

### PR DESCRIPTION
**Pull request by mzikherman**
This can probably be merged/iterated on as it's behind a feature flag.

Screen Shot 2023-10-02 at 6 04 44 PM

It works!

### CODE REVIEW

File: src/Components/SavedSearchAlert/Mutations/createAdvisoryOpportunity.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+          }
+        }
+      `,
+      variables: {
+        input: {
+          ...attributes,
+        },
+      },
+      onCompleted: response => {
+        resolve(response)
```
Comment (mzikherman) [inline src/Components/SavedSearchAlert/Mutations/createAdvisoryOpportunity.ts:39]: Will add the better error handling, where the response is introspected to see if there's any errors, and if so `reject`

File: src/Components/SavedSearchAlert/SavedSearchAlertModal.tsx [hunk 2]
```diff
@@ -178,15 +197,47 @@ export const SavedSearchAlertModal: FC<SavedSearchAlertFormProps> = ({
...
+                      </Text>
+                      <Checkbox
+                        onSelect={selected => {
+                          setFieldValue("hearFromArtsyAdvisor", selected)
+                        }}
+                        selected={values.hearFromArtsyAdvisor}
+                      />
+                    </Box>
+
+                    {values.hearFromArtsyAdvisor && (
```
Comment (mzikherman) [inline src/Components/SavedSearchAlert/SavedSearchAlertModal.tsx:228]: Leaving out the phone number stuff for now.

This may _even_ be ok to start rolling out _without_ the phone number - as it may still generate a valid email to advisory with the collector's information + email address - and the phone number could be an addition!

File: src/Components/SavedSearchAlert/types.ts [hunk 3]
```diff
@@ -63,6 +63,11 @@ export interface SavedSearchAlertFormValues {
   frequency: SavedSearchFrequency
 }
 
+export interface HearFromArtsyAdvisorFormValues {
```
Comment (mzikherman) [inline src/Components/SavedSearchAlert/types.ts:66]: I decided to separate these values rather than add them to the existing `SavedSearchAlertFormValues` interface, which is referenced in multiple other places.

File: src/Components/SavedSearchAlert/Mutations/createAdvisoryOpportunity.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  createAdvisoryOpportunityMutation,
+} from "__generated__/createAdvisoryOpportunityMutation.graphql"
+import { commitMutation, Environment, graphql } from "relay-runtime"
+
+interface MutationInput {
+  searchCriteriaID: string
+  message: string
+}
+
+export const createAdvisoryOpportunity = (
```
Comment (damassi) [inline src/Components/SavedSearchAlert/Mutations/createAdvisoryOpportunity.ts:12]: Can we use our canonical useMutation pattern as seen all about?

Comment (mzikherman) [reply src/Components/SavedSearchAlert/Mutations/createAdvisoryOpportunity.ts:12]: For this one I copied the existing `createSavedSearch` mutation as this one is linked to it - https://github.com/artsy/force/blob/f3079c38eb02009ceb4bb73f92f18bfde52782c8/src/Components/SavedSearchAlert/Mutations/createSavedSearchAlert.ts

That's how that one was setup!

File: src/Components/SavedSearchAlert/Mutations/createAdvisoryOpportunity.ts (hunk 1 above)
Comment (mzikherman) [reply src/Components/SavedSearchAlert/Mutations/createAdvisoryOpportunity.ts:41]: or rather, will update to use the preferred `useMutation` hook

**Comment by mzikherman** [comment]
Ok...feature-creeped this with the phone number selection, and it works!

It is initially set to your user's phone number w/ country code (if set).

The 'Save Alert' button is disabled unless you have at least one contactable setting checked, and additionally if the 'Hear from Artsy' box is checked- it performs the client-side phone number validation and will disable the button unless the number is valid (and it cannot be blank either).

Basically, checking the 'Hear from Artsy' box requires a non-blank valid phone number for the button to be enabled, but having an invalid or blank phone number and _not_ checking that box (and checking one or both of the other 'email' and 'push' contactable settings) will indeed enable the button. I think this is what we want 😅 

Screen Shot 2023-10-02 at 10 23 14 PM

File: src/Components/SavedSearchAlert/SavedSearchAlertModal.tsx [hunk 5]
```diff
@@ -65,6 +135,10 @@ export const SavedSearchAlertModal: FC<SavedSearchAlertFormProps> = ({
   const isFallbackToGeneratedAlertNamesEnabled = useFeatureFlag(
     "onyx_force-fallback-to-generated-alert-names"
   )
+  const isHearFromArtsyAdvisorEnabled = useFeatureFlag(
+    "onyx_advisory-opportunity-in-saved-search"
+  )
+  const { phone, regionCode } = useUserPhoneNumber()
```
Comment (mzikherman) [inline src/Components/SavedSearchAlert/SavedSearchAlertModal.tsx:141]: There were several approaches that are possible here.

One is that this modal, or possibly just the advisory opportunity / phone number part, should become a fragment container. However, that requires querying fields on `me`, and including that fragment everywhere this button is rendered.

Another approach is to take this modal or a sub-component of the advisory opportunity / phone number form, and make it a query renderer (so it can fetch the user phone information) on its own. That _might_ be more conventional than the approach taken in this PR - I'm not strongly opinionated, although the goal is to make using this shared modal (kicked off by a 'Create Alert' button) easy anywhere. Perhaps that's possible with a query renderer? Can try that or as an iteration.

The approach in this PR to try to minimize updating any callers was allowing this component to fetch the user's phone number (and country code) and inject it into the modal.

File: src/Components/SavedSearchAlert/SavedSearchAlertModal.tsx [hunk 6]
```diff
@@ -97,16 +173,53 @@ export const SavedSearchAlertModal: FC<SavedSearchAlertFormProps> = ({
...
+      }}
+      enableReinitialize
       onSubmit={handleSubmit}
+      validationSchema={Yup.object().shape({
+        phoneNumber: Yup.string()
+          .test({
+            name: "phone-number-is-valid",
+            message: "Please enter a valid phone number",
+            test: (national, context) => {
+              if (!context.parent.hearFromArtsyAdvisor) return true
```
Comment (mzikherman) [inline src/Components/SavedSearchAlert/SavedSearchAlertModal.tsx:208]: If the advisor checkbox isn't checked, we can always 'validate' this field.

Otherwise, it cannot be blank, and must be valid based on the client-side check to MP.

File: src/Components/SavedSearchAlert/Mutations/createAdvisoryOpportunity.ts (hunk 4 above)
Comment (damassi) [reply src/Components/SavedSearchAlert/Mutations/createAdvisoryOpportunity.ts:14]: Yah, this one! similar to forque: https://github.com/artsy/force/blob/main/src/Apps/Settings/Routes/Shipping/useAddAddress.ts

Comment (mzikherman) [reply src/Components/SavedSearchAlert/Mutations/createAdvisoryOpportunity.ts]: Yup - was good and added error handling!

**Comment by mzikherman** [comment]
Gonna merge this in so we can QA and iterate further!

I think I got all the corner cases - users with blank or invalid phone numbers checking the box (requiring a passing client-side validation to submit) and users where the client-side validation passes but the stricter server-side Gravity one fails.
